hazelcast/src/main/java/com/hazelcast/concurrent/reliableidgen/ReliableIdGeneratorProxy.java

```diff
@@ -0,0 +1,219 @@
+import com.hazelcast.core.Member;
+import com.hazelcast.internal.util.ThreadLocalRandomProvider;
+import com.hazelcast.logging.ILogger;
+import com.hazelcast.spi.AbstractDistributedObject;
+import com.hazelcast.spi.InternalCompletableFuture;
+import com.hazelcast.spi.NodeEngine;
+import com.hazelcast.util.Clock;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicLong;
+
+import static com.hazelcast.util.Preconditions.checkPositive;
+import static java.util.Collections.newSetFromMap;
+
+public class ReliableIdGeneratorProxy
+        extends AbstractDistributedObject<ReliableIdGeneratorService>
+        implements ReliableIdGenerator {
+
+    public static final int BITS_TIMESTAMP = 42;
+    public static final int BITS_SEQUENCE = 6;
+    public static final int BITS_NODE_ID = 16;
+
+    public static final long INCREMENT = 1 << BITS_NODE_ID;
+
+    /**
+     * 1.1.2017 0:00 GMT will be MIN_VALUE in the 42-bit signed integer.
+     * <p>
+     * {@code 1483228800000} is the value {@code System.currentTimeMillis()} would return on
+     * 1.1.2017 0:00 GMT.
+     */
+    @SuppressWarnings("checkstyle:magicnumber")
+    static final long EPOCH_START = 1483228800000L + (1L << (BITS_TIMESTAMP - 1));
+
+    private static final int NODE_ID_NOT_YET_SET = -1;
+    private static final int NODE_ID_OUT_OF_RANGE = -2;
+
+    private volatile int nodeId = NODE_ID_NOT_YET_SET;
+    private final String name;
+    private final ILogger logger;
+
+    /**
+     * The highest timestamp|seq value returned so far. The value is not shifted to most significant bits.
+     */
+    private final AtomicLong generatedValue = new AtomicLong(Long.MIN_VALUE);
+
+    private volatile Member randomMember;
+    private AutoBatcher batcher;
+
+    /**
+     * Set of member UUIDs of which we know have node IDs out of range. These members are never again used
+     * to generate unique IDs, because this error is unrecoverable.
+     */
+    private final Set<String> outOfRangeMembers = newSetFromMap(new ConcurrentHashMap<String, Boolean>());
+
+    ReliableIdGeneratorProxy(String name, NodeEngine nodeEngine, ReliableIdGeneratorService service) {
+        super(nodeEngine, service);
+        this.name = name;
+        this.logger = nodeEngine.getLogger(getClass());
+
+        ReliableIdGeneratorConfig config = nodeEngine.getConfig().findReliableIdGeneratorConfig(getName());
+        batcher = new AutoBatcher(config.getPrefetchCount(), config.getPrefetchValidityMillis(),
+                new AutoBatcher.IdBatchSupplier() {
+                    @Override
+                    public IdBatch newIdBatch(int batchSize) {
+                        return ReliableIdGeneratorProxy.this.newIdBatch(batchSize);
+                    }
+                });
+    }
+
+    @Override
+    public long newId() {
+        return batcher.newId();
```

Comment:
This will be just a shortcut, it is already guarded. The `newId()` call goes immediately to `newIdBatch()`, which goes to `getNodeId()` and this goes to `getMemberListJoinVersion()`, which is guarded for the cluster version.
There's not unit-test for this however...

Reply:
Good point, only objection is that this will break uniformity of user-facing API: when functionality is not available due to unexpected cluster version it is usual to `UnsupportedOperationException`. I think `getMemberListJoinVersion()` should also throw this kind of exception when it is denied due to too low cluster version. wdyt?

Reply:
Agree. I changed to UOE, and also added a comment to `newId()` if anyone ever wonders. Also documented that UOE is thrown and added `@since` tag to javadoc :)